vault/client.go

```diff
@@ -1,84 +1,167 @@
 package vault
 
 import (
+	"context"
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"log"
+	"sync"
+	"time"
 
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/hashicorp/vault-lambda-extension/config"
 	"github.com/hashicorp/vault/api"
 )
 
+// Client holds api.Client and handles state required to renew tokens and re-auth as required.
+type Client struct {
+	mtx sync.Mutex
+
+	VaultClient *api.Client
+	VaultConfig *api.Config
+
+	logger     *log.Logger
+	stsSvc     *sts.STS
+	authConfig config.AuthConfig
+
+	// Token refresh/renew data.
+	tokenExpiry    time.Time
+	tokenTTL       time.Duration
+	tokenRenewable bool
```

Comment:
nit: I wonder if it would be cleaner to just store the `*api.Secret` representing the token. All of these fields should be there, I think.

Reply:
Surprisingly, getting the token TTL from an `*api.Secret` can return an error, so extracting it early helps us to fail nice and early if there will be a problem. Also, even if we stored the `*api.Secret`, we would still have to have a separate field to mark the time it was fetched, because the TTL is given as a duration, not a time. Overall I think this is a bit "six of one, half a dozen of the other".